{"expand":"renderedFields,names,schema,operations,editmeta,changelog,versionedRepresentations","id":"12384082","self":"https://issues.apache.org/jira/rest/api/2/issue/12384082","key":"OPENJPA-462","fields":{"fixVersions":[{"self":"https://issues.apache.org/jira/rest/api/2/version/12313102","id":"12313102","description":"","name":"1.2.0","archived":false,"released":true,"releaseDate":"2008-08-12"}],"resolution":{"self":"https://issues.apache.org/jira/rest/api/2/resolution/1","id":"1","description":"A fix for this issue is checked into the tree and tested.","name":"Fixed"},"customfield_12312322":null,"customfield_12312323":null,"customfield_12312320":null,"customfield_12310420":"160783","customfield_12312321":null,"customfield_12312328":null,"customfield_12312329":null,"customfield_12312326":null,"customfield_12310300":null,"customfield_12312327":null,"customfield_12312324":null,"customfield_12312720":null,"customfield_12312325":null,"lastViewed":null,"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/3","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/major.svg","name":"Major","id":"3"},"labels":[],"customfield_12312333":null,"customfield_12312334":null,"customfield_12313422":"false","customfield_12312331":null,"customfield_12312332":null,"aggregatetimeoriginalestimate":null,"timeestimate":null,"customfield_12312330":null,"versions":[{"self":"https://issues.apache.org/jira/rest/api/2/version/12312340","id":"12312340","description":"","name":"0.9.7","archived":true,"released":true,"releaseDate":"2007-04-27"}],"customfield_12311120":null,"customfield_12313826":null,"customfield_12312339":null,"issuelinks":[{"id":"12320671","self":"https://issues.apache.org/jira/rest/api/2/issueLink/12320671","type":{"id":"12310010","name":"Incorporates","inward":"is part of","outward":"incorporates","self":"https://issues.apache.org/jira/rest/api/2/issueLinkType/12310010"},"inwardIssue":{"id":"12383824","key":"OPENJPA-458","self":"https://issues.apache.org/jira/rest/api/2/issue/12383824","fields":{"summary":"OpenJPA doesn't throw standard JPA exceptions","status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are not closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Done"}},"priority":{"self":"https://issues.apache.org/jira/rest/api/2/priority/2","iconUrl":"https://issues.apache.org/jira/images/icons/priorities/critical.svg","name":"Critical","id":"2"},"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/secure/viewavatar?size=xsmall&avatarId=21133&avatarType=issuetype","name":"Bug","subtask":false,"avatarId":21133}}}}],"customfield_12313825":null,"assignee":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"customfield_12312337":null,"customfield_12313823":null,"customfield_12312338":null,"customfield_12311920":null,"customfield_12313822":null,"customfield_12312335":null,"customfield_12313821":null,"customfield_12312336":null,"customfield_12313820":null,"status":{"self":"https://issues.apache.org/jira/rest/api/2/status/6","description":"The issue is considered finished, the resolution is correct. Issues which are not closed can be reopened.","iconUrl":"https://issues.apache.org/jira/images/icons/statuses/closed.png","name":"Closed","id":"6","statusCategory":{"self":"https://issues.apache.org/jira/rest/api/2/statuscategory/3","id":3,"key":"done","colorName":"green","name":"Done"}},"components":[],"customfield_12312026":null,"customfield_12312023":null,"customfield_12312024":null,"aggregatetimeestimate":null,"customfield_12312022":null,"customfield_12310921":null,"customfield_12310920":"203290","customfield_12312823":null,"creator":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ajay662","name":"ajay662","key":"ajay662","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ajay Aggarwal","active":true,"timeZone":"Etc/UTC"},"subtasks":[],"reporter":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ajay662","name":"ajay662","key":"ajay662","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Ajay Aggarwal","active":true,"timeZone":"Etc/UTC"},"aggregateprogress":{"progress":0,"total":0},"customfield_12313520":null,"customfield_12310250":null,"progress":{"progress":0,"total":0},"customfield_12313924":null,"votes":{"self":"https://issues.apache.org/jira/rest/api/2/issue/OPENJPA-462/votes","votes":0,"hasVoted":false},"worklog":{"startAt":0,"maxResults":20,"total":0,"worklogs":[]},"customfield_12313920":null,"issuetype":{"self":"https://issues.apache.org/jira/rest/api/2/issuetype/1","id":"1","description":"A problem which impairs or prevents the functions of the product.","iconUrl":"https://issues.apache.org/jira/secure/viewavatar?size=xsmall&avatarId=21133&avatarType=issuetype","name":"Bug","subtask":false,"avatarId":21133},"timespent":null,"customfield_12314020":"{summaryBean=com.atlassian.jira.plugin.devstatus.rest.SummaryBean@498d9092[summary={pullrequest=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@607326e9[overall=PullRequestOverallBean{stateCount=0, state='OPEN', details=PullRequestOverallDetails{openCount=0, mergedCount=0, declinedCount=0}},byInstanceType={}], build=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@75ee0429[overall=com.atlassian.jira.plugin.devstatus.summary.beans.BuildOverallBean@cd33013[failedBuildCount=0,successfulBuildCount=0,unknownBuildCount=0,count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], review=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@7ada576f[overall=com.atlassian.jira.plugin.devstatus.summary.beans.ReviewsOverallBean@3a652898[stateCount=0,state=<null>,dueDate=<null>,overDue=false,count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], deployment-environment=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@3f2fbcb1[overall=com.atlassian.jira.plugin.devstatus.summary.beans.DeploymentOverallBean@2a6940ef[topEnvironments=[],showProjects=false,successfulCount=0,count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], repository=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@2a00be2[overall=com.atlassian.jira.plugin.devstatus.summary.beans.CommitOverallBean@1112d2c6[count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}], branch=com.atlassian.jira.plugin.devstatus.rest.SummaryItemBean@539a2e61[overall=com.atlassian.jira.plugin.devstatus.summary.beans.BranchOverallBean@339472c5[count=0,lastUpdated=<null>,lastUpdatedTimestamp=<null>],byInstanceType={}]},errors=[],configErrors=[]], devSummaryJson={\"cachedValue\":{\"errors\":[],\"configErrors\":[],\"summary\":{\"pullrequest\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"stateCount\":0,\"state\":\"OPEN\",\"details\":{\"openCount\":0,\"mergedCount\":0,\"declinedCount\":0,\"total\":0},\"open\":true},\"byInstanceType\":{}},\"build\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"failedBuildCount\":0,\"successfulBuildCount\":0,\"unknownBuildCount\":0},\"byInstanceType\":{}},\"review\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"stateCount\":0,\"state\":null,\"dueDate\":null,\"overDue\":false,\"completed\":false},\"byInstanceType\":{}},\"deployment-environment\":{\"overall\":{\"count\":0,\"lastUpdated\":null,\"topEnvironments\":[],\"showProjects\":false,\"successfulCount\":0},\"byInstanceType\":{}},\"repository\":{\"overall\":{\"count\":0,\"lastUpdated\":null},\"byInstanceType\":{}},\"branch\":{\"overall\":{\"count\":0,\"lastUpdated\":null},\"byInstanceType\":{}}}},\"isStale\":false}}","customfield_12314141":null,"customfield_12314140":null,"project":{"self":"https://issues.apache.org/jira/rest/api/2/project/12310351","id":"12310351","key":"OPENJPA","name":"OpenJPA","projectTypeKey":"software","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/projectavatar?pid=12310351&avatarId=10043","24x24":"https://issues.apache.org/jira/secure/projectavatar?size=small&pid=12310351&avatarId=10043","16x16":"https://issues.apache.org/jira/secure/projectavatar?size=xsmall&pid=12310351&avatarId=10043","32x32":"https://issues.apache.org/jira/secure/projectavatar?size=medium&pid=12310351&avatarId=10043"},"projectCategory":{"self":"https://issues.apache.org/jira/rest/api/2/projectCategory/10252","id":"10252","description":"","name":"OpenJPA"}},"aggregatetimespent":null,"customfield_12312520":null,"customfield_12314422":null,"customfield_12314421":null,"customfield_12314146":null,"customfield_12314420":null,"customfield_12314145":null,"customfield_12314144":null,"customfield_12314143":null,"resolutiondate":"2008-07-16T02:37:34.368+0000","workratio":-1,"customfield_12312923":null,"customfield_12312920":null,"customfield_12312921":null,"watches":{"self":"https://issues.apache.org/jira/rest/api/2/issue/OPENJPA-462/watchers","watchCount":0,"isWatching":false},"created":"2007-12-06T14:56:33.317+0000","updated":"2008-08-02T14:02:41.709+0000","timeoriginalestimate":null,"description":"In my application, i wanted to catch \"javax.persistence.OptimisticLockException\" and take appropriate action. However I am seeing that instead \"org.apache.openjpa.util.OptimisticException\" is being thrown. I don't want my code to become dependent on OpenJPA and would like to see OptimisticLockException thrown.\n\nWhile discussing in the OpenJPA forums, Patrick suggested that I create a JIRA issue for this since its a bug.\nYou can see the discussion in the forum under title \"OptimisticException question\" few days ago.\n\nBelow is the stack trace from my application.\n\n<Nov 30, 2007 9:50:53 AM EST> <Error> <EJB> <BEA-010026> <Exception occurred during commit of transaction Name=[EJB com.covergence.soa.covapi.sesscfg.SessionConfigBean.getSessionConfig(com.covergence.soa.utils.SerJavaSipMessageType)],Xid=BEA1-00001267875BE126CEB7(64273527),Status=Rolled back. [Reason=<2|false|0.9.7> org.apache.openjpa.util.OptimisticException: Optimistic locking errors were detected when flushing to the data store.  The following objects may have been concurrently modified in another transaction: [com.covergence.soa.covapi.userpolicy.User-com.covergence.soa.covapi.userpolicy.User-1]],numRepliesOwedMe=0,numRepliesOwedOthers=0,seconds since begin=2,seconds left=30,SCInfo[ajayProduction+AdminServer]=(state=rolledback),properties=({weblogic.transaction.name=[EJB com.covergence.soa.covapi.sesscfg.SessionConfigBean.getSessionConfig(com.covergence.soa.utils.SerJavaSipMessageType)]}),OwnerTransactionManager=ServerTM[ServerCoordinatorDescriptor=(CoordinatorURL=AdminServer+172.30.0.202:7001+ajayProduction+t3+admin+7001+,XAResources={WLStore_ajayProduction__WLS_AdminServer},NonXAResources={})],CoordinatorURL=AdminServer+172.30.0.202:7001+ajayProduction+t3+admin+7001+): weblogic.transaction.RollbackException: Optimistic locking errors were detected when flushing to the data store.  The following objects may have been concurrently modified in another transaction: [com.covergence.soa.covapi.userpolicy.User-com.covergence.soa.covapi.userpolicy.User-1]\n\n        at weblogic.transaction.internal.TransactionImpl.throwRollbackException(TransactionImpl.java:1818)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.internalCommit(ServerTransactionImpl.java:333)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.commit(ServerTransactionImpl.java:227)\n\n        at weblogic.ejb.container.internal.BaseRemoteObject.postInvoke1(BaseRemoteObject.java:606)\n\n        at weblogic.ejb.container.internal.StatelessRemoteObject.postInvoke1(StatelessRemoteObject.java:57)\n\n        at weblogic.ejb.container.internal.BaseRemoteObject.postInvokeTxRetry(BaseRemoteObject.java:426)\n\n        at com.covergence.soa.covapi.sesscfg.SessionConfigBean_4w7egw_SessionConfigImpl.getSessionConfig(SessionConfigBean_4w7egw_SessionConfigImpl.java:76)\n\n        at jrockit.reflect.VirtualNativeMethodInvoker.invoke(Ljava.lang.Object;[Ljava.lang.Object;)Ljava.lang.Object;(Unknown Source)\n\n        at java.lang.reflect.Method.invoke(Ljava.lang.Object;[Ljava.lang.Object;I)Ljava.lang.Object;(Unknown Source)\n\n        at weblogic.ejb.container.internal.RemoteBusinessIntfProxy.invoke(RemoteBusinessIntfProxy.java:63)\n\n        at $Proxy67.getSessionConfig(Lcom.covergence.soa.utils.SerJavaSipMessageType;)Lcom.covergence.ws.callouts.SessionConfigType;(Unknown Source)\n\n        at com.covergence.soa.ws.server.CallOutsImpl.getSessionConfigType(CallOutsImpl.java:46)\n\n        at com.covergence.soa.ws.server.CallOutsImplBase.getSessionPolicy(CallOutsImplBase.java:238)\n\n        at jrockit.reflect.VirtualNativeMethodInvoker.invoke(Ljava.lang.Object;[Ljava.lang.Object;)Ljava.lang.Object;(Unknown Source)\n\n        at java.lang.reflect.Method.invoke(Ljava.lang.Object;[Ljava.lang.Object;I)Ljava.lang.Object;(Unknown Source)\n\n        at weblogic.wsee.jaxws.WLSInvoker.invoke(WLSInvoker.java:50)\n\n        at weblogic.wsee.jaxws.WLSInvoker.invoke(WLSInvoker.java:42)\n\n        at com.sun.xml.ws.server.sei.EndpointMethodHandler.invoke(EndpointMethodHandler.java:247)\n\n        at com.sun.xml.ws.server.sei.SEIInvokerPipe.process(SEIInvokerPipe.java:97)\n\n        at weblogic.wsee.jaxws.MonitoringPipe.process(MonitoringPipe.java:98)\n\n        at com.sun.xml.ws.protocol.soap.ServerMUPipe.process(ServerMUPipe.java:62)\n\n        at com.sun.xml.ws.server.WSEndpointImpl$1.process(WSEndpointImpl.java:139)\n\n        at com.sun.xml.ws.transport.http.HttpAdapter$HttpToolkit.handle(HttpAdapter.java:153)\n\n        at com.sun.xml.ws.transport.http.HttpAdapter.handle(HttpAdapter.java:235)\n\n        at com.sun.xml.ws.transport.http.servlet.ServletAdapter.handle(ServletAdapter.java:97)\n\n        at weblogic.wsee.jaxws.HttpServletAdapter.post(HttpServletAdapter.java:36)\n\n.>\n\njavax.ejb.EJBException: nested exception is: <2|false|0.9.7> org.apache.openjpa.util.OptimisticException: Optimistic locking errors were detected when flushing\n\nto the data store.  The following objects may have been concurrently modified in another transaction: [com.covergence.soa.covapi.userpolicy.User-com.covergence.\n\nsoa.covapi.userpolicy.User-1]<2|false|0.9.7> org.apache.openjpa.util.OptimisticException: Optimistic locking errors were detected when flushing to the data store.  The following objects may have been concurrently modified in another transaction: [com.covergence.soa.covapi.userpolicy.User-com.covergence.soa.covapi.userpolicy.User-1]\n\n        at org.apache.openjpa.kernel.BrokerImpl.newFlushException(BrokerImpl.java:2120)\n\n        at org.apache.openjpa.kernel.BrokerImpl.flush(BrokerImpl.java:1970)\n\n        at org.apache.openjpa.kernel.BrokerImpl.flushSafe(BrokerImpl.java:1868)\n\n        at org.apache.openjpa.kernel.BrokerImpl.beforeCompletion(BrokerImpl.java:1786)\n\n        at weblogic.transaction.internal.ServerSCInfo.doBeforeCompletion(ServerSCInfo.java:1212)\n\n        at weblogic.transaction.internal.ServerSCInfo.callBeforeCompletions(ServerSCInfo.java:1190)\n\n        at weblogic.transaction.internal.ServerSCInfo.startPrePrepareAndChain(ServerSCInfo.java:118)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.localPrePrepareAndChain(ServerTransactionImpl.java:1299)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.globalPrePrepare(ServerTransactionImpl.java:2111)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.internalCommit(ServerTransactionImpl.java:260)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.commit(ServerTransactionImpl.java:227)\n\n        at weblogic.ejb.container.internal.BaseRemoteObject.postInvoke1(BaseRemoteObject.java:606)\n\n        at weblogic.ejb.container.internal.StatelessRemoteObject.postInvoke1(StatelessRemoteObject.java:57)\n\n        at weblogic.ejb.container.internal.BaseRemoteObject.postInvokeTxRetry(BaseRemoteObject.java:426)\n\n        at com.covergence.soa.covapi.sesscfg.SessionConfigBean_4w7egw_SessionConfigImpl.getSessionConfig(SessionConfigBean_4w7egw_SessionConfigImpl.java:76)\n\n        at jrockit.reflect.VirtualNativeMethodInvoker.invoke(Ljava.lang.Object;[Ljava.lang.Object;)Ljava.lang.Object;(Unknown Source)\n\n        at java.lang.reflect.Method.invoke(Ljava.lang.Object;[Ljava.lang.Object;I)Ljava.lang.Object;(Unknown Source)\n\n        at weblogic.ejb.container.internal.RemoteBusinessIntfProxy.invoke(RemoteBusinessIntfProxy.java:63)\n\n        at $Proxy67.getSessionConfig(Lcom.covergence.soa.utils.SerJavaSipMessageType;)Lcom.covergence.ws.callouts.SessionConfigType;(Unknown Source)\n\n        at com.covergence.soa.ws.server.CallOutsImpl.getSessionConfigType(CallOutsImpl.java:46)\n\nCaused by: <2|false|0.9.7> org.apache.openjpa.util.OptimisticException: An optimistic lock violation was detected when flushing object instance \"com.covergence.soa.covapi.userpolicy.User-com.covergence.soa.covapi.userpolicy.User-1\" to the data store.  This indicates that the object was concurrently modified in anothertransaction.\n\nFailedObject: com.covergence.soa.covapi.userpolicy.User-com.covergence.soa.covapi.userpolicy.User-1\n\n        at kodo.jdbc.kernel.BatchingPreparedStatementManager.checkUpdate(BatchingPreparedStatementManager.java:354)\n\n        at kodo.jdbc.kernel.BatchingPreparedStatementManager.flushInternal(BatchingPreparedStatementManager.java:208)\n\n        at kodo.jdbc.kernel.BatchingPreparedStatementManager.flush(BatchingPreparedStatementManager.java:188)\n\n        at kodo.jdbc.kernel.ConstraintUpdateManager.flush(ConstraintUpdateManager.java:90)\n\n        at org.apache.openjpa.jdbc.kernel.AbstractUpdateManager.flush(AbstractUpdateManager.java:86)\n\n        at org.apache.openjpa.jdbc.kernel.AbstractUpdateManager.flush(AbstractUpdateManager.java:69)\n\n        at org.apache.openjpa.jdbc.kernel.JDBCStoreManager.flush(JDBCStoreManager.java:511)\n\n        at org.apache.openjpa.kernel.DelegatingStoreManager.flush(DelegatingStoreManager.java:127)\n\n        at org.apache.openjpa.datacache.DataCacheStoreManager.flush(DataCacheStoreManager.java:506)\n\n        at org.apache.openjpa.kernel.DelegatingStoreManager.flush(DelegatingStoreManager.java:127)\n\n        at org.apache.openjpa.kernel.BrokerImpl.flush(BrokerImpl.java:1970)\n\n        at org.apache.openjpa.kernel.BrokerImpl.flushSafe(BrokerImpl.java:1868)\n\n        at org.apache.openjpa.kernel.BrokerImpl.beforeCompletion(BrokerImpl.java:1786)\n\n        at weblogic.transaction.internal.ServerSCInfo.doBeforeCompletion(ServerSCInfo.java:1212)\n\n        at weblogic.transaction.internal.ServerSCInfo.callBeforeCompletions(ServerSCInfo.java:1190)\n\n        at weblogic.transaction.internal.ServerSCInfo.startPrePrepareAndChain(ServerSCInfo.java:118)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.localPrePrepareAndChain(ServerTransactionImpl.java:1299)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.globalPrePrepare(ServerTransactionImpl.java:2111)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.internalCommit(ServerTransactionImpl.java:260)\n\n        at weblogic.transaction.internal.ServerTransactionImpl.commit(ServerTransactionImpl.java:227)\n\n        at weblogic.ejb.container.internal.BaseRemoteObject.postInvoke1(BaseRemoteObject.java:606)\n\n        at weblogic.ejb.container.internal.StatelessRemoteObject.postInvoke1(StatelessRemoteObject.java:57)\n\n        at weblogic.ejb.container.internal.BaseRemoteObject.postInvokeTxRetry(BaseRemoteObject.java:426)\n\njavax.ejb.EJBException: nested exception is: <2|false|0.9.7> org.apache.openjpa.util.OptimisticException: Optimistic locking errors were detected when flushing to the data store.  The following objects may have been concurrently modified in another transaction: [com.covergence.soa.covapi.userpolicy.User-com.covergence.soa.covapi.userpolicy.User-1]\n\n        at weblogic.ejb.container.internal.RemoteBusinessIntfProxy.invoke(RemoteBusinessIntfProxy.java:78)\n\n        at $Proxy67.getSessionConfig(Lcom.covergence.soa.utils.SerJavaSipMessageType;)Lcom.covergence.ws.callouts.SessionConfigType;(Unknown Source)\n\n        at com.covergence.soa.ws.server.CallOutsImpl.getSessionConfigType(CallOutsImpl.java:46)\n\n        at com.covergence.soa.ws.server.CallOutsImplBase.getSessionPolicy(CallOutsImplBase.java:238)\n\n        at jrockit.reflect.VirtualNativeMethodInvoker.invoke(Ljava.lang.Object;[Ljava.lang.Object;)Ljava.lang.Object;(Unknown Source)\n\n        at java.lang.reflect.Method.invoke(Ljava.lang.Object;[Ljava.lang.Object;I)Ljava.lang.Object;(Unknown Source)\n\n        at weblogic.wsee.jaxws.WLSInvoker.invoke(WLSInvoker.java:50)\n\n        at weblogic.wsee.jaxws.WLSInvoker.invoke(WLSInvoker.java:42)\n\n        at com.sun.xml.ws.server.sei.EndpointMethodHandler.invoke(EndpointMethodHandler.java:247)\n\n        at com.sun.xml.ws.server.sei.SEIInvokerPipe.process(SEIInvokerPipe.java:97)\n\n        at weblogic.wsee.jaxws.MonitoringPipe.process(MonitoringPipe.java:98)\n\n        at com.sun.xml.ws.protocol.soap.ServerMUPipe.process(ServerMUPipe.java:62)\n\n        at com.sun.xml.ws.server.WSEndpointImpl$1.process(WSEndpointImpl.java:139)\n\n        at com.sun.xml.ws.transport.http.HttpAdapter$HttpToolkit.handle(HttpAdapter.java:153)\n\n        at com.sun.xml.ws.transport.http.HttpAdapter.handle(HttpAdapter.java:235)\n\n        at com.sun.xml.ws.transport.http.servlet.ServletAdapter.handle(ServletAdapter.java:97)\n\n        at weblogic.wsee.jaxws.HttpServletAdapter.post(HttpServletAdapter.java:36)\n\n        at weblogic.wsee.jaxws.JAXWSServlet.doPost(JAXWSServlet.java:218)\n\n        at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)\n\n        at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)\n\n        at weblogic.servlet.internal.StubSecurityHelper$ServletServiceAction.run(StubSecurityHelper.java:226)\n\n        at weblogic.servlet.internal.StubSecurityHelper.invokeServlet(StubSecurityHelper.java:124)\n\n        at weblogic.servlet.internal.ServletStubImpl.execute(ServletStubImpl.java:283)\n\n        at weblogic.servlet.internal.ServletStubImpl.execute(ServletStubImpl.java:175)\n\n        at weblogic.servlet.internal.WebAppServletContext$ServletInvocationAction.run(WebAppServletContext.java:3370)\n\n        at weblogic.security.acl.internal.AuthenticatedSubject.doAs(AuthenticatedSubject.java:321)\n\n        at weblogic.security.service.SecurityManager.runAs(Lweblogic.security.acl.internal.AuthenticatedSubject;Lweblogic.security.acl.internal.AuthenticatedSubject;Ljava.security.PrivilegedAction;)Ljava.lang.Object;(Unknown Source)\n\n        at weblogic.servlet.internal.WebAppServletContext.securedExecute(WebAppServletContext.java:2117)\n\n        at weblogic.servlet.internal.WebAppServletContext.execute(WebAppServletContext.java:2023)\n\n","customfield_10010":null,"timetracking":{},"customfield_12314523":null,"customfield_12314127":null,"customfield_12314522":null,"customfield_12314126":null,"customfield_12314521":null,"customfield_12314125":null,"customfield_12314520":null,"customfield_12314124":null,"attachment":[{"self":"https://issues.apache.org/jira/rest/api/2/attachment/12371177","id":"12371177","filename":"jira462.patch.1.txt","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"created":"2007-12-06T20:47:57.631+0000","size":6518,"mimeType":"text/plain","content":"https://issues.apache.org/jira/secure/attachment/12371177/jira462.patch.1.txt"},{"self":"https://issues.apache.org/jira/rest/api/2/attachment/12371181","id":"12371181","filename":"PObject.java","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"created":"2007-12-06T21:12:01.771+0000","size":505,"mimeType":"text/plain","content":"https://issues.apache.org/jira/secure/attachment/12371181/PObject.java"},{"self":"https://issues.apache.org/jira/rest/api/2/attachment/12371180","id":"12371180","filename":"TestOptimisticException.java","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"created":"2007-12-06T21:12:01.626+0000","size":2250,"mimeType":"text/plain","content":"https://issues.apache.org/jira/secure/attachment/12371180/TestOptimisticException.java"}],"customfield_12312340":null,"customfield_12314123":null,"customfield_12312341":null,"customfield_12312220":null,"customfield_12314122":null,"customfield_12314121":null,"customfield_12310041":null,"customfield_12314120":null,"customfield_12314129":null,"customfield_12314524":null,"customfield_12314128":null,"summary":"OptimisticException is thrown instead of OptimisticLockException","customfield_12314130":null,"customfield_12310291":null,"customfield_12310290":null,"customfield_12314138":null,"customfield_12314137":null,"environment":"BEA WebLogic Server 10","customfield_12314136":null,"customfield_12314135":null,"customfield_12311020":null,"customfield_12314134":null,"duedate":null,"customfield_12314132":null,"customfield_12314131":null,"comment":{"comments":[{"self":"https://issues.apache.org/jira/rest/api/2/issue/12384082/comment/12549180","id":"12549180","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"body":"An optimistic violation often manifests as a SQLException during implicit or explicit flush/commit operation. \nAs this SQL exception bubbles up the stack of OpenJPA layers, the exception gets translated with the original exception being nested within. \nThe OpenJPA facade translates an exception to javax.persistence.OptimisticLockException for the user application *if* the lower layer \n(i.e. spec-agnostic OpenJPA kernel) raises or translates original SQLException to a org.apache.openjpa.util.OptimisticException.\nIf the context is not sufficient to translate the exception to  org.apache.openjpa.util.OptimisticException, it gets translated to \norg.apache.openjpa.util.StoreException which is the generic type from all exceptions raised by the underlying datastore (which is JDBC datastore in this case).\n\nThe attached patch addresses this issue by narrowing down the original SQLException to a specific subtype of  StoreException based on the SQLState carried by the original SQLException. \nAs a result, the JPA-facade layer can translate the exception to a org.apache.opejpa.persistence.OptimisticLockException which is derived from intended javax.persistence.OptimisticLockException.\nBut this exception can get nested further up the exception translation chain before it reaches the user application.\n\nSo, if an application wants to detect whether the exception is caused by optimistic consistency violation, one way to do it is\n\n\tboolean isExpectedException(Throwable t, Class expected) {\n\t\tif (t == null) return false;\n\t\tif (expected.isAssignableFrom(t.getClass()))\n\t\t\t\treturn true;\n\t\tif (t.getCause()==t) return false;\n\t\treturn isExpectedException(t.getCause(), expected);\n\t}\n\n     and call it from the application code as:\n\n     try {\n        // do some JPA stuff;\n     } catch (Throwable t) {\n           if (isExpectedException(t, javax.persistence.OptimisticLockException.class) {\n                // do some smart thing to retry/recover\n           else\n             // rethrow\n    }\n\nNotes on the patch:\n  1. The existing DBDictionary.newStoreException() method did detect referential integrity violations by analyzing java.sql.SQLException.getSQLState(). \n  2. This patch generalizes the same logic to other violations such as locking or duplicate key.\n  3. The only context available to narrow down the cause is  java.sql.SQLException.getSQLState() -- which is ideally be same across JDBC drivers and different databases. \nBut ideals are seldom met. So a new method\n                   public String[] DBDictionary.getKnownErrorCodes(int storeExceptionType) \nis introduced, that can be overwritten by database-specific DBDictionary derivations.\n  4. The patch only supplies two error codes for referential integrity and locking errors. \n===============================================================================================\nIf someone can point me to an enumeration of these (supposedly database-neutral) error codes returned by \nSQLException.getSQLState() that will be of great help.\n===============================================================================================","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"created":"2007-12-06T20:45:39.089+0000","updated":"2007-12-06T20:45:39.089+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12384082/comment/12549182","id":"12549182","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"body":"This patch narrows down StoreException by SQLException.getSQLState().\nThis narrowing down helps the upper layers to translate the exception to more specific errors for the user.\n\nSee the previous comment for further details.\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"created":"2007-12-06T20:47:57.724+0000","updated":"2007-12-06T20:47:57.724+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12384082/comment/12549193","id":"12549193","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"body":"a test case to verify exact type of exception being thrown for optimistic consistency violation.\n\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"created":"2007-12-06T21:12:01.788+0000","updated":"2007-12-06T21:12:01.788+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12384082/comment/12550262","id":"12550262","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=pcl","name":"pcl","key":"pcl","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Patrick Linskey","active":true,"timeZone":"America/Los_Angeles"},"body":"Based on the description, it sounds like this patch might still allow internal Exception types to make it out of the OpenJPA tier unwrapped. It would seem that we should always be re-wrapping with a facade-specific exception, even if the particular exception type is not being properly handled. Is that covered in the patches at this time?","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=pcl","name":"pcl","key":"pcl","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Patrick Linskey","active":true,"timeZone":"America/Los_Angeles"},"created":"2007-12-11T01:29:29.845+0000","updated":"2007-12-11T01:29:29.845+0000"},{"self":"https://issues.apache.org/jira/rest/api/2/issue/12384082/comment/12550296","id":"12550296","author":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"body":"This patch only narrows down StoreException to more specific subtypes. So that the upper layers can translate/rewrap the exception in a more targeted exception for the user facacde.\nThis patch does not cover all internal exception types.  \nThe narrowing down is based on SQLState that the core SQLException carries with it. Hence only specific SQL errors can be narrowed within the available context/scope.\nThe user-visible exception translation will continue to happen through the layers above as The invocation context can only determine the correct translation. \n\nSpecifically, a SQL lock time out exception is now being narrowed to a internal LockException as opposed to a generic StoreException. That helps JPA facade (PersistenceExceptions) to\ntranslate it to OptimisticLockException rather than a more generic PersistenceException.    \n\nSQLExceptions are being specialized further in JDBC 4.0 and SQLExceptions.narrow() that has been introduced in this patch can be a place for more improved logic. \n\n","updateAuthor":{"self":"https://issues.apache.org/jira/rest/api/2/user?username=ppoddar%40apache.org","name":"ppoddar@apache.org","key":"ppoddar@apache.org","avatarUrls":{"48x48":"https://issues.apache.org/jira/secure/useravatar?avatarId=10452","24x24":"https://issues.apache.org/jira/secure/useravatar?size=small&avatarId=10452","16x16":"https://issues.apache.org/jira/secure/useravatar?size=xsmall&avatarId=10452","32x32":"https://issues.apache.org/jira/secure/useravatar?size=medium&avatarId=10452"},"displayName":"Pinaki Poddar","active":true,"timeZone":"America/Los_Angeles"},"created":"2007-12-11T03:59:51.929+0000","updated":"2007-12-11T03:59:51.929+0000"}],"maxResults":5,"total":5,"startAt":0},"customfield_12311820":"0|i0z63j:","customfield_12314139":null}}

